### ISSUE TITLE
[python-package] How do I reproduce LightGBM's huber loss with a custom objective?

### ISSUE THREAD

**Issue by Tahhu** [lightgbm-org/LightGBM#6041]
Hello.
I'm having some issues reproducing huber loss.

First I trained the model with a default huber loss and got results below :

``` 
# read dataset
df = pd.read_csv('data/AirPassengers.csv')
df['Month'] = pd.to_datetime(df['Month'])

# create some features
def create_calendar_vars(df, date_col):
    df_ = df.copy()
    df_['year'] = df_[date_col].dt.year.astype('int')
    df_['quarter'] = df_[date_col].dt.quarter.astype('int')
    df_['month'] = df_[date_col].dt.month.astype('int')
    return df_

df = create_calendar_vars(df, date_col='Month')

# train set
train = df[df['Month'] < pd.to_datetime('1958-01-01')].copy()
train['split'] = 'train'

# val set
val = df[(df['Month'] >= pd.to_datetime('1958-01-01'))
         & (df['Month'] < pd.to_datetime('1960-01-01'))
         ].copy()
val['split'] = 'val'

# test set
test = df[df['Month'] >= pd.to_datetime('1960-01-01')].copy()
test['split'] = 'test'

# all sets
df = pd.concat((train, val, test))

# train model
train_ds = lgbm.Dataset(train[feature2train],  train[target_col])
val_ds = lgbm.Dataset(val[feature2train],  val[target_col], reference=train_ds)
eval_set = [train_ds, val_ds]

params_b = {
    "n_jobs": 10,
    "random_state": 42,
    "early_stopping": 10,
    "learning_rate": 0.05,
    "boost_from_average": True,
    "max_depth": -1,
    "num_threads": 8,
    "verbose": -1,
    "objective": "huber",
    "alpha": 0.9
}

model = lgbm.train(
                params_b,
                train_ds,
                num_boost_round=500,
                valid_sets=eval_set,
                verbose_eval=100,
                  )

df['predict'] = model.predict(df[feature2train]) 
```

Actually, my predict is a constant - 252.3630, But it is ok here.

![image](https://github.com/microsoft/LightGBM/assets/16069272/bccf11b3-ab46-4f1f-bca2-749da4a08eb2)

Then I  carefully copied the huber loss function and its gradients from [documentation](https://github.com/microsoft/LightGBM/blob/e9fbd19d7cbaeaea1ca54a091b160868fc5c79ec/src/objective/regression_objective.hpp#L307)

``` 
@dataclasses.dataclass()
class HuberLossFunction:
    """
    Loss (x)
        = 1/2 * x^2, |x| < alpha
        = alpha (|x| - 1/2 * alpha), |x| >= alpha
        = alpha * (|x| - 1/2 * alpha ** 2), |x| >= alpha
        
    Gradient (x)
        = x, |x| < alpha
        = alpha * sign(x), |x| >= alpha

    """
    alpha: float = dataclasses.field(default=0.9)

    def gradient_hessian(self, preds, y_true, weight):
        residual = (preds - y_true).astype('float')
        gradient = np.where(np.abs(residual) <= self.alpha, residual, np.sign(residual) * self.alpha)
        hessian = np.ones(preds.shape)

        if weight is not None:
            return gradient * weight, hessian * weight
        else:
            return gradient, hessian

    def __call__(self, preds, train_data):
        return self.gradient_hessian(preds, train_data.get_label(), train_data.get_weight())
   
def huber_custom_eval(preds, data):
    y_true = data.get_label()
    y_pred = preds
    residual = (y_pred - y_true).astype("float")
    alpha = .9
    loss = np.where(np.abs(residual) <= alpha , .5 * ((residual) ** 2), alpha * np.abs(residual) - .5 * (alpha ** 2))
    
    return "huber_custom", np.mean(loss), False
```

And trained the model again with custom loss and metric:

``` 
train_ds = lgbm.Dataset(train[feature2train], 
                        train[target_col],
                       )

val_ds = lgbm.Dataset(val[feature2train], 
                       val[target_col],
                       reference=train_ds
                       )

test_ds = lgbm.Dataset(test[feature2train], 
                       test[target_col],
                       reference=train_ds
                       )

eval_set = [train_ds, val_ds]

params_b = {
    "n_jobs": 10,
    "random_state": 42,
    "early_stopping": 10,
    "learning_rate": 0.05,
    "boost_from_average": True,
    "max_depth": -1,
    "num_threads": 8,
    "verbose": -1,
}

model = lgbm.train(
                params_b,
                train_ds,
                num_boost_round=500,
                valid_sets=eval_set,
                verbose_eval=100,
                fobj= huber_custom_loss,
                feval=huber_custom_eval
                  )

df['predict'] = model.predict(df[feature2train]) 
```

I got zero predictions. 
Predictions doesn't math. I didn't get correct results.

Could you suggest what I'm missing? Why results doesn't match? 
Also read the #3532 , but it doesn't help me =(

![image](https://github.com/microsoft/LightGBM/assets/16069272/4053509c-d544-4432-b2ce-f1c01d8136ad)
